### PR TITLE
[Frontend][Core] Standardize request error handling with VLLMError hierarchy

### PR DESCRIPTION
**Pull request by zqzten**
## Purpose

This PR implements Steps 1–4 of RFC #48227:

- Introduce `VLLMError` / `VLLMClientError` / `VLLMServerError` and reparent existing client/server exceptions under this hierarchy.
- Migrate engine/request validation `ValueError`s in `sampling_params`, `pooling_params`, `input_processor`, and `async_llm` to `VLLMValidationError`.
- Catch `VLLMClientError` (instead of bare `ValueError`) in `AsyncLLM.generate` / pooling so client errors are not wrapped as `EngineGenerateError` (500).
- Register a single `app.exception_handler(VLLMError)` and dispatch HTTP status codes from the exception hierarchy in `create_error_response`, thus simplify API server exception handling.
- Also migrate enc-dec `prompt_embeds` rejection in `inputs/engine.py` (was incorrectly becoming 500 on the generate path).
- Convert custom logits processor `validate_params` failures (`ValueError`) to `VLLMValidationError` in `validate_logits_processors_parameters` for backward compatibility for legacy custom logits processors.

### Related works

#49217 / #49214 migrate individual entrypoint validators to `VLLMValidationError` which can be regarded as part of Step 5 of RFC #48227. This PR establishes the shared hierarchy and engine/API dispatch that those changes can rely on.

### Backward compatibility

This has been covered in RFC:

> - **API behavior**: No change for correctly-typed errors. The only behavior change is that hypothetical `ValueError` from internal code paths (model executor, attention layers) would become 500 instead of 400 — this is a bug fix, not a regression.
> - **`VLLMValidationError` / `VLLMUnprocessableEntityError` inheritance**: Changing from `ValueError` to `VLLMClientError` breaks code that catches `ValueError` expecting to catch `VLLMValidationError` / `VLLMUnprocessableEntityError`. External code users who depend on this inheritance are unlikely but would need to update their catch clauses.

### Other notes

- After some investigation, I decide not to introduce a separate 413 error for context-length exceeded but keep it as 400 for OpenAI-compatible behavior.

## Test Plan

```bash
pytest \
  tests/entrypoints/serve/instrumentator/test_http_status_metrics.py \
  tests/tool_use/test_chat_completion_request_validations.py \
  tests/tool_use/test_responses_request_validations.py \
  tests/entrypoints/openai/responses/test_sampling_params.py::TestResponsesRequestSamplingParams::test_structured_outputs_and_json_schema_conflict \
  tests/entrypoints/openai/completion/test_prompt_validation.py::test_disable_prompt_embeds \
  tests/renderers/test_chat_utils_prompt_embeds.py::test_parse_chat_messages_requires_flag \
  tests/renderers/test_chat_utils_prompt_embeds.py::test_parse_chat_messages_rejects_missing_data \
  tests/test_sampling_params.py \
  tests/test_pooling_params.py::test_removed_pooling_parameters \
  tests/v1/structured_output/test_validation.py \
  tests/samplers/test_non_finite_params.py \
  tests/multimodal/media/test_unprocessable_entity_error.py \
  -v
```

## Test Result

All above regression tests passed.

### CODE REVIEW

File: tests/engine/test_short_mm_context.py [hunk 1]
```diff
@@ -19,7 +21,9 @@
 def test_context_length_too_short(vllm_runner, image_assets, model):
     images = [asset.pil_image for asset in image_assets]
 
-    with pytest.raises(ValueError, match="longer than the maximum model length"):
+    with pytest.raises(
+        VLLMValidationError, match="longer than the maximum model length"
+    ):
         vllm_model = vllm_runner(
             model,
```
Comment (noooop) [inline tests/engine/test_short_mm_context.py:28]: ValueError, as a basic Python exception, is used in many places. Is it really necessary to change all of them to VLLMValidationError?

@DarkLight1337

Comment (DarkLight1337) [reply tests/engine/test_short_mm_context.py:28]: I think using specific errors this makes error handling easier

**Comment by noooop** [comment]
awesome!

**Comment by euisuh** [comment]
Thanks for pushing the error hierarchy cleanup — the direction makes sense. One thing I’d be careful about: this PR deletes `tests/entrypoints/serve/instrumentator/test_http_status_metrics.py`, which was covering the regression where Prometheus `http_requests_total` could record client-side validation failures as `5xx` even though the HTTP response is `4xx`. Since this PR changes the exception-handler registration and status-code dispatch path, I think we should keep equivalent coverage somewhere rather than removing it outright. Otherwise a future middleware/handler ordering change could regress metrics while normal HTTP response tests still pass.

**Comment by zqzten** [comment]
> Thanks for pushing the error hierarchy cleanup — the direction makes sense. One thing I’d be careful about: this PR deletes `tests/entrypoints/serve/instrumentator/test_http_status_metrics.py`, which was covering the regression where Prometheus `http_requests_total` could record client-side validation failures as `5xx` even though the HTTP response is `4xx`. Since this PR changes the exception-handler registration and status-code dispatch path, I think we should keep equivalent coverage somewhere rather than removing it outright. Otherwise a future middleware/handler ordering change could regress metrics while normal HTTP response tests still pass.

Yeah actually the previous UT is not very good because it does regression test by copy the impl rather than directly call the actual impl so that it does not really serve as a regression guard for that case. I'll try to figure out how to make a UT which is practically effective for it.

**Comment by zqzten** [comment]
It seems like there're some error type mismatch in test, let me fix it.

**Comment by zqzten** [comment]
> > Thanks for pushing the error hierarchy cleanup — the direction makes sense. One thing I’d be careful about: this PR deletes `tests/entrypoints/serve/instrumentator/test_http_status_metrics.py`, which was covering the regression where Prometheus `http_requests_total` could record client-side validation failures as `5xx` even though the HTTP response is `4xx`. Since this PR changes the exception-handler registration and status-code dispatch path, I think we should keep equivalent coverage somewhere rather than removing it outright. Otherwise a future middleware/handler ordering change could regress metrics while normal HTTP response tests still pass.
> 
> Yeah actually the previous UT is not very good because it does regression test by copy the impl rather than directly call the actual impl so that it does not really serve as a regression guard for that case. I'll try to figure out how to make a UT which is practically effective for it.

I've rewritten this test to call the actual impl, now it shall work as expected.

**Comment by umut-polat** [comment]
Nice to see this land as the canonical Step 2 of #48227. One suggestion for the `sampling_params.py` migration: consider passing `parameter=`/`value=` on the `VLLMValidationError`s (e.g. `parameter="temperature", value=self.temperature`), not just the type swap. The RFC's own motivation notes those structured fields surface as the `param` field in the OpenAI error response, so populating them here is where the client-facing benefit actually shows up. I had this in #35664 (which I'm closing in favor of this) — happy to help wire the kwargs across the `_verify_args` sites if useful.

**Comment by zqzten** [comment]
> Nice to see this land as the canonical Step 2 of #48227. One suggestion for the `sampling_params.py` migration: consider passing `parameter=`/`value=` on the `VLLMValidationError`s (e.g. `parameter="temperature", value=self.temperature`), not just the type swap. The RFC's own motivation notes those structured fields surface as the `param` field in the OpenAI error response, so populating them here is where the client-facing benefit actually shows up. I had this in #35664 (which I'm closing in favor of this) — happy to help wire the kwargs across the `_verify_args` sites if useful.

That's a good point. I think you can still keep the original PR because that won't conflict with this one. This PR focus on the hierarchy refactor, then separate following up PRs can gradually improve the readability of `VLLMValidationError`, not limited to `sampling_params.py` here, thus make the change trajectory cleaner. Thanks for help with the error handling anyway!

**Review by noooop — APPROVED** [review]
